### PR TITLE
Adam optimizer consistent with paper

### PR DESCRIPTION
**Pull request by formath**

### CODE REVIEW

**Comment by sergeykolychev** [comment]
@formath please see https://github.com/formath/mxnet/pull/1

**Comment by formath** [comment]
@sergeykolychev Have been merged.
@piiswrong

**Comment by piiswrong** [comment]
@sxjscience Could you verify this against the paper?

@formath The default for rho should be 1.0 to keep the original default behavior

**Comment by ZiyueHuang** [comment]
@dsqx71  Seems the problem we discuss before. I remember that adam now in master is better according to your experiments but is indeed different from paper. Am I correct? I'm not sure.

**Comment by formath** [comment]
@piiswrong I have set the default `rho` 1.0. It could have the same result as the origin master but also give a chance for users to set different hyper-parameter on different data.

File: src/operator/optimizer_op-inl.h [hunk 1]
```diff
@@ -632,6 +632,7 @@ struct AdamParam : public dmlc::Parameter<AdamParam> {
   float beta1;
   float beta2;
   float epsilon;
+  float rho;
```
Comment (piiswrong) [inline src/operator/optimizer_op-inl.h]: it looks like rho doesn't need to be in backend? You can do the shinking in frontend.

Comment (formath) [reply src/operator/optimizer_op-inl.h]: Yes. But `rho` is also a hyper-parameter. it should be added into `struct AdamParam` like that of `lr`. `lr` is shrinking in frontend but also have a part in backend.

**Comment by sxjscience** [comment]
Why do we need rho? I feel that this parameter is not useful.

**Comment by formath** [comment]
In my use, namely training `field aware factorization machine` on 20 millions features and billions instances, setting `rho` 0.999 really works. It could get an absolute 0.01 promotion on AUC. I don't know whether it works on cv or nlp tasks but giving user a chance could be useful. @sxjscience

**Comment by sxjscience** [comment]
@formath I feel that the `rho` has the effect to gradually transforms the gradient estimator from biased to unbiased, which may have some advantages. However, I've checked the Adam paper and haven't found the rho hyper-parameter. Could you help point out the section in the paper? Also, we need to better document the it's usage.

**Comment by formath** [comment]
@sxjscience `rho` is the `lambda` in the paper

**Comment by formath** [comment]
@sxjscience `rho` is `lambda` in the paper
```
Kingma, Diederik, and Jimmy Ba. “Adam: A method for stochastic optimization.” arXiv preprint arXiv:1412.6980 (2014).
```

Also, I have tested FTML and fount it works better than FTRL

**Comment by sxjscience** [comment]
@formath I see. I've checked again different versions of the Adam paper and find the rho in the v2 and v3 versions: https://arxiv.org/pdf/1412.6980v2.pdf, https://arxiv.org/pdf/1412.6980v2.pdf. However, it's somehow removed in the latest arxiv version (v9), https://arxiv.org/pdf/1412.6980v9.pdf .

**Comment by piiswrong** [comment]
If it's removed from the paper and no other frameworks support it I don't think we should add it

**Comment by formath** [comment]
Closed it now. Consider this later if needed. Thanks.
